models/Document.php

```diff
@@ -865,49 +866,56 @@ public function getFullPath(bool $force = false)
         // the hardlink there are snippets embedded and this snippets have links pointing to a document which is also
         // inside the hardlink scope, but this is an ID link, so we cannot rewrite the link the usual way because in the
         // snippet / link we don't know anymore that whe a inside a hardlink wrapped document
-        if (!$link && \Pimcore\Tool::isFrontend() && Site::isSiteRequest() && !FrontendTool::isDocumentInCurrentSite($this)) {
-            if ($mainRequest && ($mainDocument = $mainRequest->get(DynamicRouter::CONTENT_KEY))) {
-                if ($mainDocument instanceof WrapperInterface) {
-                    $hardlinkPath = '';
-                    $hardlink = $mainDocument->getHardLinkSource();
-                    $hardlinkTarget = $hardlink->getSourceDocument();
-
-                    if ($hardlinkTarget) {
-                        $hardlinkPath = preg_replace('@^' . preg_quote(Site::getCurrentSite()->getRootPath(), '@') . '@', '', $hardlink->getRealFullPath());
-
-                        $link = preg_replace('@^' . preg_quote($hardlinkTarget->getRealFullPath(), '@') . '@',
-                            $hardlinkPath, $this->getRealFullPath());
-                    }
+        if (!$link && Tool::isFrontend()) {
+            $differentDomain = false;
+            $site = FrontendTool::getSiteForDocument($this);
+            if ($site instanceof Site){
+                $differentDomain = $site->getMainDomain() != $request->getHost();
+            }
+
+            if ((Site::isSiteRequest() && !FrontendTool::isDocumentInCurrentSite($this))
+                || (Tool::isFrontendRequestByAdmin() && $differentDomain)){
```

Comment:
@dvesh3 do you mean like this? 

```suggestion
            if (Tool::isFrontendRequestByAdmin() && $site instanceof Site){
                $differentDomain = $site->getMainDomain() != $request->getHost();
            }

            if ((Site::isSiteRequest() && !FrontendTool::isDocumentInCurrentSite($this))
                || $differentDomain){
```